common/testing/testlogger/testlogger.go

```diff
@@ -0,0 +1,501 @@
+			logExpectations: false,
+			level:           zapcore.DebugLevel,
+			logCaller:       true,
+		},
+	}
+	tl.state.failOnError.Store(true)
+	tl.state.failOnDPanic.Store(true)
+	tl.state.mu.expectations = map[Level]*list.List{
+		Debug: list.New(),
+		Info:  list.New(),
+		Warn:  list.New(),
+		Error: list.New(),
+	}
+	for _, opt := range opts {
+		opt(tl)
+	}
+	if tl.wrapped == nil {
+		ztl := zaptest.NewLogger(t,
+			zaptest.Level(tl.state.level),
+			zaptest.WrapOptions(
+				zap.AddStacktrace(zap.ErrorLevel),
+				zap.WithCaller(tl.state.logCaller),
+			),
+		)
+		tl.wrapped = log.NewZapLogger(ztl).Skip(1)
+	}
+
+	// Only possible with a *testing.T until *rapid.T supports `Cleanup`
+	if ct, ok := t.(CleanupCapableT); ok {
+		// NOTE(tim): We don't care about anything logged after the test completes. Sure, this is racy,
+		// but it reduces the likelihood that we see stupid errors due to testing.T.Logf race conditions...
+		ct.Cleanup(tl.Close)
+	}
+
+	return tl
+}
+
+// ResetUnexpectedErrors resets the unexpected error state, returning the previous value.
+func (tl *TestLogger) ResetUnexpectedErrors() bool {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	prevUnexpectedLogs := tl.state.mu.unexpectedErrs
+	tl.state.mu.unexpectedErrs = false
+	return prevUnexpectedLogs
+}
+
+// Expect instructs the logger to expect certain errors, as specified by the msg and tag arguments.
+// This is useful for ignoring certain errors that are expected.
+func (tl *TestLogger) Expect(level Level, msg string, tags ...tag.Tag) *Expectation {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	if tl.state.logExpectations {
+		tl.wrapped.Info(fmt.Sprintf("(%p) TestLogger::Expecting: '%s'\n", tl, msg))
+	}
+	return tl.addExpectationLocked(false, level, msg, tags...)
+}
+
+// DontExpect instructs the logger to _not_ expect certain errors, as specified by the msg and tag arguments.
+// This is useful for ensuring that certain errors _do not_ occur during a test.
+func (tl *TestLogger) DontExpect(level Level, msg string, tags ...tag.Tag) *Expectation {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	if tl.state.logExpectations {
+		tl.wrapped.Info(fmt.Sprintf("(%p) TestLogger::NotExpecting: '%s'\n", tl, msg))
+	}
+	return tl.addExpectationLocked(true, level, msg, tags...)
+}
+
+func (tl *TestLogger) addExpectationLocked(unexpected bool, level Level, msg string, tags ...tag.Tag) *Expectation {
+	e := &Expectation{
+		testLogger: tl,
+		lvl:        level,
+		unexpected: unexpected,
+	}
+	m := newMatcher(msg, tags, e)
+	e.e = tl.state.mu.expectations[level].PushBack(m)
+	return e
+}
+
+// Forget removes a previously registered expectation.
+// It will no longer be used to determine if a log message is expected or unexpected.
+func (tl *TestLogger) Forget(e *Expectation) {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	tl.state.mu.expectations[e.lvl].Remove(e.e)
+}
+
+func (tl *TestLogger) isUnexpected(level Level, msg string, tags []tag.Tag) bool {
```

Comment:
I'm not a fan of `isUnexpected` as the new name for this (the original is `matchesExpectation`), as that combined with your new bool is confusing.

How about `shouldFailTest(...)`? That's clear about whats expected regardless of the mode this testlogger is in